### PR TITLE
feat: Publish NIC topology matrix and GPU NUMA nodes in metadata file

### LINKED ISSUE #1211: [Feature]: Add NIC state detection logic + Metadata collector changes
### Prerequisites

- [x] I searched existing issues

### Feature Summary

Implement logic for NIC state detection + the changes in metadata collector as outlined in this design doc: https://github.com/KaivalyaMDabhadkar/NVSentinel/blob/642-nic-monitor-design/docs/nic-health-monitor/link-state-detection.md

### Problem/Use Case

Implement logic for NIC state detection + the changes in metadata collector as outlined in this design doc: https://github.com/KaivalyaMDabhadkar/NVSentinel/blob/642-nic-monitor-design/docs/nic-health-monitor/link-state-detection.md

### Proposed Solution

Refer to the design doc: Implement logic for NIC state detection + the changes in metadata collector as outlined in this design doc: https://github.com/KaivalyaMDabhadkar/NVSentinel/blob/642-nic-monitor-design/docs/nic-health-monitor/link-state-detection.md

### Component

Health Monitor

### PR DESCRIPTION
**Pull request by KaivalyaMDabhadkar**
## Summary

The NIC Health Monitor needs to classify each NIC as Compute, Storage, or Management to avoid false `REPLACE_VM` remediations on management NICs. This classification requires two pieces of information that only the GPU driver can provide:

1. **GPU-NIC topology relationships** (PIX/PXB/NODE/SYS) : determines whether a NIC is on the compute fabric (shares a PCIe switch with a GPU) or a storage/general NIC
2. **Per-GPU NUMA affinity** : determines which CPU sockets host GPUs, so NICs on non-GPU sockets can be excluded as management

Previously, neither was available in the metadata file. The NIC monitor would have had to invoke `nvidia-smi` itself (adding a GPU driver runtime dependency) or walk sysfs PCIe paths (fragile, platform-specific). This PR makes the metadata collector the single source of truth for topology data: it runs `nvidia-smi topo -m` once at startup, parses the output, and publishes the raw matrix and GPU NUMA nodes in `gpu_metadata.json`. The NIC monitor then reads the JSON and performs all classification locally - no `nvidia-smi` dependency, no sysfs PCIe heuristics.

**Metadata Collector Changes**

- **New `nvidia-smi topo -m` parser** (`pkg/nic/topo_parser.go`): Parses the GPU-NIC relationship matrix from `nvidia-smi topo -m` output into a structured `TopoMatrix`. Handles multiple real-world output formats:
  - Direct `mlx5_N` column names (DGX H100)
  - `NICn` column labels with a `NIC Legend` remapping block (OCI H100 / older drivers)
  - Non-`mlx5` device names (e.g., `ibp3s0`, `roceP6p3s0` on GB200)
  - ANSI escape codes in piped output (stripped before parsing)
  - Grace/GB200 comma-separated NUMA Affinity format (`"0,2-17"` -> CPU socket node `0`)

- **Data model additions** (`data-models/pkg/model/gpu_metadata.go`):
  - `NICTopology map[string][]string` on `GPUMetadata` - raw per-NIC topology levels (one entry per GPU, in `gpus[]` order)
  - `NUMANode int` on `GPUInfo` - GPU's CPU-socket NUMA affinity, parsed from the `nvidia-smi topo -m` NUMA Affinity column

- **Collector wiring** (`pkg/collector/collector.go`): After NVML collection, runs the topo parser and populates `nic_topology` and `gpus[].numa_node` in the published `gpu_metadata.json`. Logs a warning if the GPU count from the topo matrix doesn't match NVML.

**Design Principle: Responsibility Split**

The metadata collector publishes **raw facts** only - the verbatim topology-level matrix and the GPU NUMA node. It does not classify NICs or make monitoring decisions. All interpretation (Compute vs Storage vs Management, NUMA gating, HCA-based DPU exclusion) lives in the NIC Health Monitor. This means classification rules can evolve without changing the collector, and the same metadata file can serve future consumers with different classification needs.

**Manual Testing**

Deployed on A100, H100, GB200 and L40S clusters and tested and verified that the metadata file gets correctly populated with the right topology information.

## Type of Change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Core Services
- [ ] Documentation/CI
- [ ] Fault Management
- [x] Health Monitors
- [ ] Janitor
- [ ] Other: ____________

## Testing
- [x] Tests pass locally
- [x] Manual testing completed
- [x] No breaking changes (or documented)

## Checklist
- [x] Self-review completed
- [x] Documentation updated (if needed)
- [x] Ready for review

## Summary by CodeRabbit

* **New Features**
  * GPU metadata now includes NIC↔GPU topology and per‑GPU NUMA affinity. Topology collection is integrated into the metadata collector and will populate these fields when available; collection errors or GPU/NIC count mismatches are non‑fatal and leave fields unset.

* **Tests**
  * Extensive unit tests added covering topology detection, parsing, collector integration, and many NVIDIA output formats and edge cases.

### CODE REVIEW

File: metadata-collector/pkg/nic/topo_parser.go [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+	}
+
+	if countGPUTokens(accumulated) < minGPUTokens {
+		return 0, nil, fmt.Errorf(
+			"no header row with multiple GPU columns found in nvidia-smi topo -m output (found %d)",
+			countGPUTokens(accumulated),
+		)
+	}
+
+	return lastIdx, accumulated, nil
```
[the remark this answers is not in the document]
Comment (KaivalyaMDabhadkar) [reply metadata-collector/pkg/nic/topo_parser.go:350]: I have added TestParseTopoMatrix_SingleGPUIsNotMistakenForHeader explicitly testing this. Single-GPU nodes don't benefit from NIC role classification (there's nothing to compare against for homogeneity).

File: metadata-collector/pkg/collector/collector.go [hunk 2]
```diff
@@ -91,9 +117,86 @@ func (c *Collector) Collect(ctx context.Context) (*model.GPUMetadata, error) {
...
+		slog.Warn("Failed to collect NIC topology matrix, nic_topology will be empty",
+			"error", err)
+
+		return
+	}
+
+	if matrix == nil || len(matrix.NICs) == 0 {
+		slog.Info("No InfiniBand NICs reported by nvidia-smi topo -m, nic_topology will be empty")
+		return
+	}
```
[the remark this answers is not in the document]
Comment (KaivalyaMDabhadkar) [reply metadata-collector/pkg/collector/collector.go:146]: Addressed

File: metadata-collector/pkg/collector/collector.go [hunk 3]
```diff
@@ -91,9 +117,86 @@ func (c *Collector) Collect(ctx context.Context) (*model.GPUMetadata, error) {
...
+// as we can in order.
+func populateGPUNUMANodes(metadata *model.GPUMetadata, matrix *nic.TopoMatrix) {
+	n := len(metadata.GPUs)
+	if len(matrix.GPUNUMANodes) < n {
+		n = len(matrix.GPUNUMANodes)
+	}
+
+	for i := range n {
+		metadata.GPUs[i].NUMANode = matrix.GPUNUMANodes[i]
+	}
```
[the remark this answers is not in the document]
Comment (KaivalyaMDabhadkar) [reply metadata-collector/pkg/collector/collector.go]: Addressed

File: metadata-collector/pkg/collector/collector.go [hunk 4]
```diff
@@ -91,9 +117,87 @@ func (c *Collector) Collect(ctx context.Context) (*model.GPUMetadata, error) {
...
+		slog.Warn("Failed to collect NIC topology matrix, nic_topology will be empty",
+			"error", err)
+
+		return
+	}
+
+	if matrix == nil || len(matrix.NICs) == 0 {
+		slog.Info("No InfiniBand NICs reported by nvidia-smi topo -m, nic_topology will be empty")
+		return
+	}
```
[the remark this answers is not in the document]
Comment (KaivalyaMDabhadkar) [reply metadata-collector/pkg/collector/collector.go]: Addressed

File: metadata-collector/pkg/collector/collector.go [hunk 5]
```diff
@@ -91,9 +117,92 @@ func (c *Collector) Collect(ctx context.Context) (*model.GPUMetadata, error) {
...
+
 	return metadata, nil
 }
 
+// populateNICTopology attaches the raw nvidia-smi topo -m matrix to
+// metadata.NICTopology. Failures degrade to an empty map rather than
+// aborting the collector — other consumers of gpu_metadata.json still
+// receive the NVML-derived fields.
+func (c *Collector) populateNICTopology(ctx context.Context, metadata *model.GPUMetadata) {
+	if c.nicTopo == nil {
```
Comment (lalitadithya) [inline metadata-collector/pkg/collector/collector.go]: under what cases would this be nil? can we instead check upstream and bail out of this code path early?

File: metadata-collector/pkg/collector/collector.go [hunk 6]
```diff
@@ -91,9 +117,92 @@ func (c *Collector) Collect(ctx context.Context) (*model.GPUMetadata, error) {
...
+			metadata.GPUs[i].NUMANode = matrix.GPUNUMANodes[i]
+		}
+	}
+}
+
+// checkGPUCountMismatch returns a short description when the parsed
+// matrix disagrees with NVML's GPU count, or an empty string when they
+// match. The caller emits a WARN so the mismatch is visible without
+// aborting the collector.
+func checkGPUCountMismatch(matrix *nic.TopoMatrix, metadata *model.GPUMetadata) string {
```
Comment (lalitadithya) [inline metadata-collector/pkg/collector/collector.go]: should this be returning an error instead?

File: metadata-collector/pkg/nic/topo_parser.go [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+// header wrapping (seen on OCI H100 pipes where GPU0 lands on its own
+// line), out-of-order/duplicated data rows, preamble noise, and both
+// NIC naming conventions.
+func ParseTopoMatrix(output string) (*TopoMatrix, error) {
+	output = ansiEscape.ReplaceAllString(output, "")
+	lines := strings.Split(output, "\n")
+
+	headerIdx, headers, err := findHeaderRow(lines)
+	if err != nil {
+		return nil, err
```
Comment (lalitadithya) [inline metadata-collector/pkg/nic/topo_parser.go]: can we wrap errors please?

File: metadata-collector/pkg/collector/collector.go (hunk 5 above)
Comment (KaivalyaMDabhadkar) [reply metadata-collector/pkg/collector/collector.go]: It was only used for the tests-path, removed it

File: metadata-collector/pkg/collector/collector.go (hunk 6 above)
Comment (KaivalyaMDabhadkar) [reply metadata-collector/pkg/collector/collector.go]: good catch, changed checkGPUCountMismatch to return error instead of string

File: metadata-collector/pkg/nic/topo_parser.go (hunk 7 above)
Comment (KaivalyaMDabhadkar) [reply metadata-collector/pkg/nic/topo_parser.go]: Done

**Review by lalitadithya — APPROVED** [review]
